### ISSUE TITLE
inserting rows into tables with `SELECT` + `LIMT 1` slower than inserting with just the `SELECT`

### ISSUE THREAD

**Issue by ruchirK** [MaterializeInc/materialize#8495]
Running from a recent main (+ some persistence feature work but this reproduces even with persistence disabled)

I ran materialize with `cargo run --release -- -w 1 --log-file=stderr --experimental` (I don't think the experimental is important here)

I set up with the following preparatory sql commands

```
materialize=> create table ten (f1 int);
CREATE TABLE
Time: 3.510 ms
materialize=> create table thousand (f1 int);
CREATE TABLE
Time: 4.419 ms
materialize=> create table to_insert(k1 int, k2 int, v text);
CREATE TABLE
Time: 3.214 ms
materialize=> create table t1 (k1 int, k2 int, v text);
CREATE TABLE
Time: 4.723 ms
materialize=> insert into ten values (0), (1), (2), (3), (4);
INSERT 0 5
Time: 0.589 ms
materialize=> insert into ten values (5), (6), (7), (8), (9);
INSERT 0 5
Time: 0.570 ms
materialize=> insert into thousand select a1.f1 * 100 + a2.f1 * 10 + a3.f1 from ten as a1, ten as a2, ten as a3;
INSERT 0 1000
Time: 6.252 ms
materialize=> insert into to_insert select a1.f1, a2.f1, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' from thousand as a1, thousand as a2;
INSERT 0 1000000
Time: 3192.385 ms (00:03.192)
```

And now, will repeatedly insert things from `to_insert` into `t1`:

```
materialize=> insert into t1 select * from to_insert;
INSERT 0 1000000
Time: 1396.605 ms (00:01.397)
materialize=> insert into t1 select * from to_insert limit 1;
INSERT 0 1
Time: 20649.209 ms (00:20.649)
```

Here, as you can see, its strange that inserting 1 row takes so much longer than inserting 1 MM rows.

The query itself is fast:

```
materialize=> select * from to_insert limit 1;
 k1 | k2 |                                                  v
----+----+------------------------------------------------------------------------------------------------------
  0 |  0 | aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
(1 row)

Time: 1.914 ms
```

When I profile Materialize while running this query it seems like we are spending a lot of time in the dataflow layer doing various kinds of trace maintenance before returning a result

<img width="1249" alt="Screen Shot 2021-10-01 at 3 56 39 PM" src="https://user-images.githubusercontent.com/4015012/135679215-2bf1cd85-2db1-43bf-a0d8-64f86130746e.png">

Maybe hypothesis: we're somehow not hitting the fast path for this kind of query when its part of an insert statement?

**Comment by wangandi** [comment]
Correct: we are not hitting the fast path for this kind of query. All inserts involve trivial `RowSetFinishing`s
https://github.com/MaterializeInc/materialize/blob/45fdf4aef2f1561c697a5afdd8d3524fcc701bef/src/coord/src/coord.rs#L3914

Looking at the `plan_insert_query` code,
https://github.com/MaterializeInc/materialize/blob/45fdf4aef2f1561c697a5afdd8d3524fcc701bef/src/sql/src/plan/query.rs#L247-L307
it looks like one reason that we are not hitting fast path is that we always put a map + project around the expression, which causes the query `select * from to_insert limit 1` to be parsed like so:
```
| Get to_insert
| TopK group=() order=() limit=1 offset=0
| Project (#0..#2)
| Map
| Project (#0..#2)
```
And thus, we do not notice that we can extract `RowSetFinishing` in this situation.

I feel that it is not hard to play around with the `plan_insert_query` code to produce a `RowSetFinishing`, but a more complete solution is to extract `RowSetFinishing` later in the optimization process (see #8678).

Also see #8194 for a different situation where we are failing to hit fast path.
